File: pkg/acpbridge/bridge.go
package acpbridge

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"log/slog"
	"strconv"
	"sync"
	"time"

	"github.com/coder/acp-go-sdk"
	"github.com/halfwhey/allagent/pkg/clients/claude"
	"github.com/halfwhey/allagent/pkg/wrap"
)

const (
	defaultPaneWidth  = 120
	defaultPaneHeight = 40

	// idleTimeout is how long the drain loop waits after the last event
	// (with no pending tool calls) before declaring the turn complete.
	idleTimeout = 3 * time.Second
)

// session tracks one ACP session backed by a Claude Code wrapper.
type session struct {
	id      string
	wrapper wrap.Wrapper
	cancel  context.CancelFunc // cancels current Prompt turn
}

// Bridge implements acp.Agent by forwarding prompts to a wrap.Wrapper
// (Claude Code in tmux) and streaming events back as ACP notifications.
type Bridge struct {
	conn     *acp.AgentSideConnection
	model    string
	sessions map[string]*session
	mu       sync.Mutex
	log      *slog.Logger
}

// Compile-time interface checks.
var _ acp.Agent = (*Bridge)(nil)

// New creates a Bridge with the given default model.
func New(model string) *Bridge {
	if model == "" {
		model = "sonnet"
	}
	return &Bridge{
		model:    model,
		sessions: make(map[string]*session),
		log:      slog.Default(),
	}
}

// SetAgentConnection stores the connection so we can send notifications.
func (b *Bridge) SetAgentConnection(conn *acp.AgentSideConnection) {
	b.conn = conn
}

// SetLogger overrides the default logger.
func (b *Bridge) SetLogger(l *slog.Logger) {
	b.log = l
}

// Shutdown stops all sessions and kills their tmux processes.
func (b *Bridge) Shutdown() {
	b.mu.Lock()
	defer b.mu.Unlock()
	n := len(b.sessions)
	for sid, s := range b.sessions {
		b.log.Info("stopping session", "sid", sid)
		_ = s.wrapper.Stop()
	}
	b.sessions = make(map[string]*session)
	b.log.Info("shutdown complete", "sessions_stopped", n)
}

// ---------------------------------------------------------------------------
// acp.Agent methods
// ---------------------------------------------------------------------------

func (b *Bridge) Initialize(_ context.Context, params acp.InitializeRequest) (acp.InitializeResponse, error) {
	clientName := ""
	if params.ClientInfo != nil {
		clientName = params.ClientInfo.Name
		if params.ClientInfo.Version != "" {
			clientName += " " + params.ClientInfo.Version
		}
	}
	b.log.Info("client connected",
		"client", clientName,
		"protocol_version", params.ProtocolVersion,
	)
	return acp.InitializeResponse{
		ProtocolVersion: acp.ProtocolVersionNumber,
		AgentCapabilities: acp.AgentCapabilities{
			LoadSession: false,
		},
		AgentInfo: &acp.Implementation{
			Name:    "allagent",
			Title:   acp.Ptr("allagent (Claude Code)"),
			Version: "0.1.0",
		},
	}, nil
}

func (b *Bridge) Authenticate(_ context.Context, _ acp.AuthenticateRequest) (acp.AuthenticateResponse, error) {
	return acp.AuthenticateResponse{}, nil
}

func (b *Bridge) NewSession(ctx context.Context, params acp.NewSessionRequest) (acp.NewSessionResponse, error) {
	sid := randomID()
	tmuxName := "allagent-" + sid[:12]

	w := claude.New(claude.Config{
		Model:       b.model,
		WorkingDir:  params.Cwd,
		TmuxSession: tmuxName,
		PaneWidth:   defaultPaneWidth,
		PaneHeight:  defaultPaneHeight,
	})
	if err := w.Start(ctx); err != nil {
		return acp.NewSessionResponse{}, fmt.Errorf("start claude: %w", err)
	}

	b.mu.Lock()
	b.sessions[sid] = &session{id: sid, wrapper: w}
	b.mu.Unlock()

	b.log.Info("session created", "sid", sid, "cwd", params.Cwd, "model", b.model)
	return acp.NewSessionResponse{SessionId: acp.SessionId(sid)}, nil
}

func (b *Bridge) Prompt(_ context.Context, params acp.PromptRequest) (acp.PromptResponse, error) {
	sid := string(params.SessionId)
	b.mu.Lock()
	s, ok := b.sessions[sid]
	b.mu.Unlock()
	if !ok {
		return acp.PromptResponse{}, fmt.Errorf("session %s not found", sid)
	}

	// Cancel any prior turn.
	b.mu.Lock()
	if s.cancel != nil {
		s.cancel()
	}
	ctx, cancel := context.WithCancel(context.Background())
	s.cancel = cancel
	b.mu.Unlock()

	// Extract text from prompt content blocks.
	text := extractPromptText(params.Prompt)
	if text == "" {
		cancel()
		return acp.PromptResponse{StopReason: acp.StopReasonEndTurn}, nil
	}

	b.log.Info("prompt received", "sid", sid, "length", len(text))

	if err := s.wrapper.Send(text); err != nil {
		cancel()
		return acp.PromptResponse{}, fmt.Errorf("send: %w", err)
	}

	reason := b.drainEvents(ctx, sid, s.wrapper)

	b.mu.Lock()
	s.cancel = nil
	b.mu.Unlock()

	b.log.Info("prompt completed", "sid", sid, "stop_reason", reason)
	return acp.PromptResponse{StopReason: reason}, nil
}

func (b *Bridge) Cancel(_ context.Context, params acp.CancelNotification) error {
	sid := string(params.SessionId)
	b.log.Info("cancel requested", "sid", sid)
	b.mu.Lock()
	s, ok := b.sessions[sid]
	b.mu.Unlock()
	if !ok {
		return nil
	}
	b.mu.Lock()
	if s.cancel != nil {
		s.cancel()
	}
	b.mu.Unlock()
	_ = s.wrapper.Interrupt()
	return nil
}

func (b *Bridge) SetSessionMode(_ context.Context, _ acp.SetSessionModeRequest) (acp.SetSessionModeResponse, error) {
	return acp.SetSessionModeResponse{}, nil
}

// ---------------------------------------------------------------------------
// Event draining
// ---------------------------------------------------------------------------

// drainEvents reads from the wrapper's event channel and sends ACP
// notifications until the turn completes or the context is cancelled.
func (b *Bridge) drainEvents(ctx context.Context, sid string, w wrap.Wrapper) acp.StopReason {
	var (
		pendingTools int
		gotResponse  bool
		idle         *time.Timer
	)

	resetIdle := func() {
		if idle != nil {
			idle.Stop()
		}
		idle = time.NewTimer(idleTimeout)
	}
	resetIdle()

	for {
		select {
		case <-ctx.Done():
			if idle != nil {
				idle.Stop()
			}
			return acp.StopReasonCancelled

		case ev, ok := <-w.Events():
			if !ok {
				// Channel closed -- backend died.
				if idle != nil {
					idle.Stop()
				}
				return acp.StopReasonEndTurn
			}
			resetIdle()
			b.log.Debug("event received", "sid", sid, "kind", ev.Kind, "tool", ev.ToolName, "text_len", len(ev.Text))
			b.sendEvent(ctx, sid, ev, &pendingTools, &gotResponse)

			if ev.Kind == wrap.EventToolCall && ev.ToolName == "AskUserQuestion" {
				b.handleUserQuestion(ctx, sid, ev, w)
			}

		case <-idle.C:
			if gotResponse && pendingTools <= 0 {
				b.log.Debug("idle timeout, ending turn", "sid", sid)
				return acp.StopReasonEndTurn
			}
			b.log.Debug("idle timeout, still waiting", "sid", sid, "pending_tools", pendingTools, "got_response", gotResponse)
			resetIdle()
		}
	}
}

// sendEvent translates a single wrap.Event into ACP notification(s).
func (b *Bridge) sendEvent(ctx context.Context, sid string, ev wrap.Event, pendingTools *int, gotResponse *bool) {
	sessionID := acp.SessionId(sid)

	var update *acp.SessionUpdate
	switch ev.Kind {
	case wrap.EventAssistantText:
		*gotResponse = true
		u := acp.UpdateAgentMessageText(ev.Text)
		update = &u

	case wrap.EventAssistantThink:
		u := acp.UpdateAgentThoughtText(ev.Text)
		update = &u

	case wrap.EventToolCall:
		*pendingTools++
		b.log.Info("tool call", "tool", ev.ToolName, "id", ev.ToolID, "pending", *pendingTools)
		kind := mapToolKind(ev.ToolName)
		locs := extractLocations(ev.ToolInput)
		var rawInput map[string]any
		_ = json.Unmarshal([]byte(ev.ToolInput), &rawInput)

		opts := []acp.ToolCallStartOpt{
			acp.WithStartKind(kind),
			acp.WithStartStatus(acp.ToolCallStatusInProgress),
		}
		if rawInput != nil {
			opts = append(opts, acp.WithStartRawInput(rawInput))
		}
		if len(locs) > 0 {
			opts = append(opts, acp.WithStartLocations(locs))
		}
		u := acp.StartToolCall(acp.ToolCallId(ev.ToolID), ev.ToolName, opts...)
		update = &u

	case wrap.EventToolResult:
		*pendingTools--
		if *pendingTools < 0 {
			*pendingTools = 0
		}
		b.log.Info("tool result", "id", ev.ToolID, "pending", *pendingTools)
		u := acp.UpdateToolCall(
			acp.ToolCallId(ev.ToolID),
			acp.WithUpdateStatus(acp.ToolCallStatusCompleted),
			acp.WithUpdateContent([]acp.ToolCallContent{
				acp.ToolContent(acp.TextBlock(ev.Text)),
			}),
		)
		update = &u

	case wrap.EventError:
		u := acp.UpdateAgentMessageText("Error: " + ev.Text)
		update = &u

	case wrap.EventSystem:
		if ev.Text == "" {
			return
		}
		u := acp.UpdateAgentThoughtText(ev.Text)
		update = &u

	default:
		// EventUserMessage, EventSessionStart, EventTurnStart, EventTurnEnd
		return
	}

	if update == nil {
		return
	}
	if err := b.conn.SessionUpdate(ctx, acp.SessionNotification{
		SessionId: sessionID,
		Update:    *update,
	}); err != nil {
		b.log.Error("failed to send session update", "err", err, "kind", ev.Kind)
	}
}

// ---------------------------------------------------------------------------
// AskUserQuestion -> ACP RequestPermission
// ---------------------------------------------------------------------------

// handleUserQuestion intercepts an AskUserQuestion tool call and forwards each
// question to the ACP client via RequestPermission. The selected option index
// is sent back to the tmux pane as a digit so Claude Code can continue.
func (b *Bridge) handleUserQuestion(ctx context.Context, sid string, ev wrap.Event, w wrap.Wrapper) {
	parsed := parseAskUserQuestion(ev.ToolInput)
	if parsed == nil {
		b.log.Warn("AskUserQuestion: could not parse input, sending default answer", "sid", sid)
		_ = w.Send("1")
		return
	}

	for _, q := range parsed.Questions {
		header := q.Header
		if header == "" {
			header = "Claude has a question"
		}

		opts := make([]acp.PermissionOption, 0, len(q.Options)+1)
		for i, opt := range q.Options {
			name := opt.Label
			if opt.Description != "" {
				name += " — " + opt.Description
			}
			opts = append(opts, acp.PermissionOption{
				Kind:     acp.PermissionOptionKindAllowOnce,
				OptionId: acp.PermissionOptionId(strconv.Itoa(i + 1)),
				Name:     name,
			})
		}
		opts = append(opts, acp.PermissionOption{
			Kind:     acp.PermissionOptionKindRejectOnce,
			OptionId: "0",
			Name:     "Cancel",
		})

		for _, opt := range opts {
			b.log.Debug("permission option", "sid", sid, "id", opt.OptionId, "name", opt.Name, "kind", opt.Kind)
		}

		content := []acp.ToolCallContent{acp.ToolContent(acp.TextBlock(q.Question))}
		kind := acp.ToolKindOther
		status := acp.ToolCallStatusInProgress

		resp, err := b.conn.RequestPermission(ctx, acp.RequestPermissionRequest{
			SessionId: acp.SessionId(sid),
			ToolCall: acp.RequestPermissionToolCall{
				ToolCallId: acp.ToolCallId(ev.ToolID),
				Title:      &header,
				Kind:       &kind,
				Status:     &status,
				Content:    content,
			},
			Options: opts,
		})

		if err != nil {
			b.log.Error("RequestPermission failed", "err", err, "sid", sid)
			_ = w.Interrupt()
			return
		}
		if resp.Outcome.Cancelled != nil {
			b.log.Info("user cancelled question", "sid", sid, "header", header)
			_ = w.Interrupt()
			return
		}
		if resp.Outcome.Selected != nil {
			id := string(resp.Outcome.Selected.OptionId)
			if id == "0" {
				b.log.Info("user selected cancel", "sid", sid, "header", header)
				_ = w.Interrupt()
				return
			}
			b.log.Info("user selected option", "sid", sid, "header", header, "optionId", id)
			if sendErr := w.Send(id); sendErr != nil {
				b.log.Error("failed to send answer to tmux", "err", sendErr)
			}
		}
	}
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

// extractPromptText concatenates all text content blocks from a prompt.
func extractPromptText(blocks []acp.ContentBlock) string {
	var text string
	for _, b := range blocks {
		if b.Text != nil {
			text += b.Text.Text
		}
	}
	return text
}

func randomID() string {
	var b [12]byte
	if _, err := rand.Read(b[:]); err != nil {
		return fmt.Sprintf("sess_%d", time.Now().UnixNano())
	}
	return "sess_" + hex.EncodeToString(b[:])
}

File: packages/claudraband-cli/src/acpbridge/bridge.ts
import type {
  Agent,
  AgentSideConnection,
  AuthenticateRequest,
  AuthenticateResponse,
  CancelNotification,
  ContentBlock,
  InitializeRequest,
  InitializeResponse,
  ListSessionsRequest,
  ListSessionsResponse,
  LoadSessionRequest,
  LoadSessionResponse,
  NewSessionRequest,
  NewSessionResponse,
  PermissionOption,
  PromptRequest,
  PromptResponse,
  SessionConfigOption,
  SessionConfigSelectOption,
  SessionMode,
  SessionModeState,
  SessionUpdate,
  SetSessionConfigOptionRequest,
  SetSessionConfigOptionResponse,
  SetSessionModeRequest,
  SetSessionModeResponse,
  ToolCallContent,
} from "@agentclientprotocol/sdk";
import { PROTOCOL_VERSION } from "@agentclientprotocol/sdk";
import {
  createClaudraband,
  EventKind,
  MODEL_OPTIONS,
  PERMISSION_MODES,
  type Claudraband,
  type ClaudrabandEvent,
  type ClaudrabandLogger,
  type ClaudrabandPermissionDecision,
  type ClaudrabandPermissionRequest,
  type ClaudrabandSession,
  type PermissionMode,
  type TerminalBackend,
  type TurnDetectionMode,
} from "claudraband-core";
import { extractLocations, mapToolKind } from "./toolmap";

function buildConfigOptions(
  model: string,
  permissionMode: PermissionMode,
): SessionConfigOption[] {
  return [
    {
      type: "select",
      id: "model",
      name: "Model",
      category: "model",
      currentValue: model,
      options: MODEL_OPTIONS as unknown as SessionConfigSelectOption[],
    } as SessionConfigOption,
    {
      type: "select",
      id: "permission_mode",
      name: "Permission Mode",
      category: "mode",
      currentValue: permissionMode,
      options: PERMISSION_MODES.map((mode) => ({
        value: mode.id,
        name: mode.name,
        description: mode.description,
      })),
    } as SessionConfigOption,
  ];
}

function buildModes(currentModeId: PermissionMode): SessionModeState {
  const availableModes: SessionMode[] = PERMISSION_MODES.map((mode) => ({
    id: mode.id,
    name: mode.name,
    description: mode.description,
  }));
  return { availableModes, currentModeId };
}

interface BridgeSession {
  session: ClaudrabandSession;
}

function makeDefaultLogger(): ClaudrabandLogger {
  return {
    info: (msg, ...args) => console.log(msg, ...args),
    debug: (msg, ...args) => console.debug(msg, ...args),
    warn: (msg, ...args) => console.warn(msg, ...args),
    error: (msg, ...args) => console.error(msg, ...args),
  };
}

export class Bridge implements Agent {
  private conn!: AgentSideConnection;
  private runtime: Claudraband;
  private defaultClaudeArgs: string[];
  private defaultModel: string;
  private defaultPermissionMode: PermissionMode;
  private defaultTerminalBackend: TerminalBackend;
  private defaultTurnDetection: TurnDetectionMode;
  private sessions = new Map<string, BridgeSession>();
  private log: ClaudrabandLogger;

  constructor(config: {
    claudeArgs: string[];
    model: string;
    permissionMode: PermissionMode;
    terminalBackend: TerminalBackend;
    turnDetection: TurnDetectionMode;
  }) {
    this.defaultClaudeArgs = config.claudeArgs;
    this.defaultModel = config.model || "sonnet";
    this.defaultPermissionMode = config.permissionMode;
    this.defaultTerminalBackend = config.terminalBackend;
    this.defaultTurnDetection = config.turnDetection;
    this.log = makeDefaultLogger();
    this.runtime = createClaudraband({
      claudeArgs: this.defaultClaudeArgs,
      model: this.defaultModel,
      permissionMode: this.defaultPermissionMode,
      terminalBackend: this.defaultTerminalBackend,
      turnDetection: this.defaultTurnDetection,
    });
  }

  setConnection(conn: AgentSideConnection): void {
    this.conn = conn;
  }

  setLogger(log: ClaudrabandLogger): void {
    this.log = log;
    this.runtime = createClaudraband({
      claudeArgs: this.defaultClaudeArgs,
      model: this.defaultModel,
      permissionMode: this.defaultPermissionMode,
      terminalBackend: this.defaultTerminalBackend,
      turnDetection: this.defaultTurnDetection,
      logger: log,
    });
  }

  shutdown(): void {
    for (const [sid, entry] of this.sessions) {
      this.log.info("detaching session", sid);
      entry.session.detach().catch(() => {});
    }
    this.sessions.clear();
  }

  async initialize(params: InitializeRequest): Promise<InitializeResponse> {
    const clientName = params.clientInfo
      ? `${params.clientInfo.name}${params.clientInfo.version ? ` ${params.clientInfo.version}` : ""}`
      : "";
    this.log.info(
      "client connected",
      "client",
      clientName,
      "protocol_version",
      params.protocolVersion,
    );

    return {
      protocolVersion: PROTOCOL_VERSION,
      agentCapabilities: {
        loadSession: true,
        sessionCapabilities: {
          list: {},
        },
      },
      agentInfo: {
        name: "claudraband",
        title: "claudraband (Claude Code)",
        version: "0.1.0",
      },
    };
  }

  async authenticate(
    _params: AuthenticateRequest,
  ): Promise<AuthenticateResponse> {
    return {};
  }

  async newSession(params: NewSessionRequest): Promise<NewSessionResponse> {
    const session = await this.runtime.openSession({
      cwd: params.cwd,
      claudeArgs: this.defaultClaudeArgs,
      model: this.defaultModel,
      permissionMode: this.defaultPermissionMode,
      logger: this.log,
      onPermissionRequest: (request: ClaudrabandPermissionRequest) =>
        this.handlePermissionRequest(request),
    });
    this.sessions.set(session.sessionId, { session });
    this.attachSession(session);

    return {
      sessionId: session.sessionId,
      configOptions: buildConfigOptions(session.model, session.permissionMode),
      modes: buildModes(session.permissionMode),
    };
  }

  async loadSession(params: LoadSessionRequest): Promise<LoadSessionResponse> {
    const history = await this.runtime.replaySession(params.sessionId, params.cwd);
    for (const event of history) {
      this.sendEvent(params.sessionId, event);
    }

    const session = await this.runtime.openSession({
      sessionId: params.sessionId,
      cwd: params.cwd,
      claudeArgs: this.defaultClaudeArgs,
      model: this.defaultModel,
      permissionMode: this.defaultPermissionMode,
      logger: this.log,
      onPermissionRequest: (request: ClaudrabandPermissionRequest) =>
        this.handlePermissionRequest(request),
    });
    this.sessions.set(params.sessionId, { session });
    this.attachSession(session);

    return {
      configOptions: buildConfigOptions(session.model, session.permissionMode),
      modes: buildModes(session.permissionMode),
    };
  }

  async listSessions(
    params: ListSessionsRequest,
  ): Promise<ListSessionsResponse> {
    const sessions = (await this.runtime.listSessions(params.cwd ?? undefined))
      .filter((session) => session.source === "live" && session.alive);
    return { sessions };
  }

  async prompt(params: PromptRequest): Promise<PromptResponse> {
    const session = this.requireSession(params.sessionId);
    return session.prompt(extractPromptText(params.prompt));
  }

  async cancel(params: CancelNotification): Promise<void> {
    const entry = this.sessions.get(params.sessionId);
    if (!entry) return;
    await entry.session.interrupt();
  }

  async setSessionMode(
    params: SetSessionModeRequest,
  ): Promise<SetSessionModeResponse> {
    const session = this.requireSession(params.sessionId);
    await session.setPermissionMode(params.modeId as PermissionMode);
    this.conn.sessionUpdate({
      sessionId: params.sessionId,
      update: {
        sessionUpdate: "current_mode_update",
        currentModeId: params.modeId,
      } as SessionUpdate,
    }).catch(() => {});
    return {};
  }

  async setSessionConfigOption(
    params: SetSessionConfigOptionRequest,
  ): Promise<SetSessionConfigOptionResponse> {
    const session = this.requireSession(params.sessionId);
    const value = String(params.value);

    switch (params.configId) {
      case "model":
        await session.setModel(value);
        break;
      case "permission_mode":
        await session.setPermissionMode(value as PermissionMode);
        this.conn.sessionUpdate({
          sessionId: params.sessionId,
          update: {
            sessionUpdate: "current_mode_update",
            currentModeId: value,
          } as SessionUpdate,
        }).catch(() => {});
        break;
      default:
        throw new Error(`unknown config option: ${params.configId}`);
    }

    return {
      configOptions: buildConfigOptions(session.model, session.permissionMode),
    };
  }

  private requireSession(sessionId: string): ClaudrabandSession {
    const entry = this.sessions.get(sessionId);
    if (!entry) {
      throw new Error(`session ${sessionId} not found`);
    }
    return entry.session;
  }

  private attachSession(session: ClaudrabandSession): void {
    (async () => {
      for await (const event of session.events()) {
        this.sendEvent(session.sessionId, event);
      }
    })().catch((err: unknown) => {
      this.log.error("session event loop failed", "sid", session.sessionId, err);
    });
  }

  private sendEvent(sessionId: string, event: ClaudrabandEvent): void {
    let update: SessionUpdate | null = null;

    switch (event.kind) {
      case EventKind.AssistantText:
        update = {
          sessionUpdate: "agent_message_chunk",
          content: { type: "text", text: event.text },
        };
        break;
      case EventKind.AssistantThinking:
      case EventKind.System:
        if (!event.text) return;
        update = {
          sessionUpdate: "agent_thought_chunk",
          content: { type: "text", text: event.text },
        };
        break;
      case EventKind.UserMessage:
        return;
      case EventKind.ToolCall: {
        let rawInput: Record<string, unknown> | undefined;
        try {
          rawInput = JSON.parse(event.toolInput) as Record<string, unknown>;
        } catch {
          rawInput = undefined;
        }

        const toolCallObj: Record<string, unknown> = {
          sessionUpdate: "tool_call",
          toolCallId: event.toolID,
          title: event.toolName,
          kind: mapToolKind(event.toolName),
          status: "pending",
        };
        if (rawInput) {
          toolCallObj.rawInput = rawInput;
        }
        const locations = extractLocations(event.toolInput);
        if (locations) {
          toolCallObj.locations = locations;
        }
        update = toolCallObj as SessionUpdate;
        break;
      }
      case EventKind.ToolResult: {
        const content: ToolCallContent[] = [
          {
            type: "content",
            content: { type: "text", text: event.text },
          } as ToolCallContent,
        ];
        update = {
          sessionUpdate: "tool_call_update",
          toolCallId: event.toolID,
          status: "completed",
          content,
        } as SessionUpdate;
        break;
      }
      case EventKind.Error:
        update = {
          sessionUpdate: "agent_message_chunk",
          content: { type: "text", text: `Error: ${event.text}` },
        };
        break;
      default:
        return;
    }

    this.conn.sessionUpdate({ sessionId, update }).catch((err: unknown) => {
      this.log.error("failed to send session update", "sid", sessionId, err);
    });
  }

  private async handlePermissionRequest(
    request: ClaudrabandPermissionRequest,
  ): Promise<ClaudrabandPermissionDecision> {
    const content: ToolCallContent[] = request.content.map((block) => ({
      type: "content",
      content: block as unknown as ContentBlock,
    })) as ToolCallContent[];

    const options: PermissionOption[] = request.options
      .filter((option) => !option.textInput)
      .map((option) => ({
      kind: option.kind,
      optionId: option.optionId,
      name: option.name,
    }));

    const response = await this.conn.requestPermission({
      sessionId: request.sessionId,
      toolCall: {
        toolCallId: request.toolCallId,
        title: request.title,
        kind: request.kind,
        status: "pending",
        content,
      },
      options,
    });

    if (response.outcome.outcome === "selected") {
      return {
        outcome: "selected",
        optionId: response.outcome.optionId,
      };
    }

    return { outcome: "cancelled" };
  }
}

function extractPromptText(blocks: ContentBlock[] | undefined): string {
  if (!blocks) return "";
  let text = "";
  for (const block of blocks) {
    if ("text" in block && typeof block.text === "string") {
      text += block.text;
    }
  }
  return text;
}
